spec/lib/signed-token-request.spec.js

```diff
@@ -0,0 +1,256 @@
+                expect(actualToken).toEqual(token);
+                expect(signer.sign).toHaveBeenCalled();
+                actualRequestOpts = requestAsyncMock.argsForCall[0][0];
+                expect(actualRequestOpts.method).toEqual(method);
+                expect(actualRequestOpts.url).toEqual(`${url}?public=true`);
+                assertAuthHeader(actualRequestOpts.headers.Authorization);
+            });
+        });
+    });
+    describe(".downloadToFile", () => {
+        it("successfuly downloads to a \"file\" from public endpoint", () => {
+            var file, promise, stream;
+
+            file = "fakeFile";
+            stream = createStream("test");
+            responseMock.pipe.andReturn(stream);
+            fsMock.createWriteStream.andReturn({
+                path: file
+            });
+            url = `${url}/${token}`;
+            promise = privateSigned.downloadToFile(token, file).then((actualFile) => {
+                expect(requestHandler.createUrl).toHaveBeenCalledWith(account.id, false, token);
+                expect(requestHandler.buildRequestOptions).toHaveBeenCalledWith(url, "GET");
+                expect(actualFile).toEqual(file);
+                expect(signer.sign).toHaveBeenCalled();
+            });
+
+            // This is horrible but I could not find a better solution to it.
+            // Have to wait for the signer.sign promise to resolve before emitting
+            // the appropriate events to complete the response.
+            setTimeout(() => {
+                requestMock.emit("response", responseMock);
+                stream.emit("finish");
+            }, 250);
+
+            return promise;
+        });
+    });
+    describe(".uploadFromFile", () => {
+        var uploadResp;
+
+        uploadResp = {
+            body: "test",
+            statusCode: 201,
+            headers: {
+                location: `${url}/${token}`
+            }
+        };
+        it("successfully uploads from a file", () => {
+            var promise, stream;
+
+            stream = createStream("Test");
+            fsMock.createReadStream = jasmine.createSpy("fs.createReadStream").andReturn(stream);
+            requestMock.on = (event, callback) => {
+                callback(uploadResp);
+            };
+            promise = privateSigned.uploadFromFile("file.json").then((actualToken) => {
+                expect(requestHandler.createUrl).toHaveBeenCalledWith(account.id, false);
+                expect(requestHandler.buildRequestOptions).toHaveBeenCalledWith(url, "POST");
+                expect(actualToken).toEqual(token);
+                expect(signer.sign).toHaveBeenCalled();
+            });
+
+            // Basically the same thing as above. Have to have this event fired after
+            // the signer has signed. I found no other way to do this than setTimeout.
+            setTimeout(() => {
```

Comment:
Above solution would be great because you can leverage whatever promise that signer.sign returns and trigger further events.